src/Uno.UI/UI/Xaml/Internal/InputManager.Pointers.Managed.cs

```diff
@@ -289,24 +279,30 @@ internal void OnPointerMoved(Windows.UI.Core.PointerEventArgs args)
 
 			if (this.Log().IsEnabled(LogLevel.Trace))
 			{
-				this.Log().Trace($"CoreWindow_PointerMoved [{originalSource.GetDebugName()}");
+				this.Log().Trace($"CoreWindow_PointerMoved [{originalSource.GetDebugName()}]");
 			}
 
 			var routedArgs = new PointerRoutedEventArgs(args, originalSource);
 
 			// First raise the PointerExited events on the stale branch
 			if (staleBranch.HasValue)
 			{
-				Raise(Leave, staleBranch.Value, routedArgs);
+				if (Raise(Leave, staleBranch.Value, routedArgs) is { VisualTreeAltered: true })
+				{
+					// The visual tree have been modified in a way that requires to perform a new hit test.
+					originalSource = HitTest(args).element ?? Windows.UI.Xaml.Window.Current.Content;
+				}
 			}
 
 			// Second (try to) raise the PointerEnter on the OriginalSource
 			// Note: This won't do anything if already over.
-			routedArgs.Handled = false;
-			Raise(Enter, originalSource, routedArgs);
+			if (Raise(Enter, originalSource, routedArgs) is { VisualTreeAltered: true })
+			{
+				// The visual tree have been modified in a way that requires to perform a new hit test.
+				originalSource = HitTest(args).element ?? Windows.UI.Xaml.Window.Current.Content;
```

Comment:
Quite concerned about performance. I see that `VisualTreeAltered` is added to reduce the frequency of having to re-call `HitTest`, but still not sure if we will still be calling HitTest too much and if it will have a negative impact.

Reply:
Yes it will have a negative perf impact, that's why I added that flag that should be `true` in few rare cases. And actually it does not cover all cases (for instance if a `RenderTransform` is updated by the enter we will stiff dispatch the move to it), we are only covering cases that could drive to be stuck into an invalid state (pressed while element is no longer in the visual tree), other cases will just be recovered on next pointer event.